apps/server/src/modules/tool/uc/external-tool.uc.spec.ts

```diff
@@ -0,0 +1,325 @@
+
+		uc = module.get(ExternalToolUc);
+		externalToolService = module.get(ExternalToolService);
+		externalToolMapper = module.get(ExternalToolRequestMapper);
+		authorizationService = module.get(AuthorizationService);
+		oauthProviderService = module.get(OauthProviderService);
+		oAuthEncryptionService = module.get(DefaultEncryptionService);
+	});
+
+	afterAll(async () => {
+		await module.close();
+		await orm.close();
+	});
+
+	afterEach(() => {
+		jest.resetAllMocks();
+	});
+
+	function setup() {
+		const basicConfig: BasicToolConfigDO = new BasicToolConfigDO({
+			type: ToolConfigType.BASIC,
+			baseUrl: 'baseUrl',
+		});
+
+		const oauth2ConfigWithoutExternalData: Oauth2ToolConfigDO = new Oauth2ToolConfigDO({
+			type: ToolConfigType.OAUTH2,
+			baseUrl: 'baseUrl',
+			clientId: 'clientId',
+			skipConsent: false,
+		});
+
+		const oauth2ConfigWithExternalData: Oauth2ToolConfigDO = new Oauth2ToolConfigDO({
+			type: ToolConfigType.OAUTH2,
+			baseUrl: 'baseUrl',
+			clientId: 'clientId',
+			clientSecret: 'clientSecret',
+			skipConsent: false,
+			tokenEndpointAuthMethod: TokenEndpointAuthMethod.CLIENT_SECRET_POST,
+			frontchannelLogoutUri: 'frontchannelLogoutUri',
+			scope: 'openid offline',
+			redirectUris: ['redirectUri1', 'redirectUri2'],
+		});
+
+		const lti11Config: Lti11ToolConfigDO = new Lti11ToolConfigDO({
+			type: ToolConfigType.LTI11,
+			baseUrl: 'baseUrl',
+			key: 'key',
+			secret: 'secret',
+			resource_link_id: 'resource_link_id',
+			lti_message_type: LtiMessageType.BASIC_LTI_LAUNCH_REQUEST,
+			privacy_permission: LtiPrivacyPermission.ANONYMOUS,
+		});
+
+		const externalToolDO: ExternalToolDO = new ExternalToolDO({
+			name: 'name',
+			url: 'url',
+			logoUrl: 'logoUrl',
+			config: basicConfig,
+			parameters: [],
+			isHidden: false,
+			openNewTab: false,
+			version: 1,
+		});
+
+		const oauthClient: ProviderOauthClient = {
+			client_id: 'clientId',
+			client_secret: 'clientSecret',
+			token_endpoint_auth_method: TokenEndpointAuthMethod.CLIENT_SECRET_POST,
+			frontchannel_logout_uri: 'frontchannelLogoutUri',
+			scope: 'openid offline',
+			redirect_uris: ['redirectUri1', 'redirectUri2'],
+		};
+
+		const user: User = userFactory.buildWithId();
+		const currentUser: ICurrentUser = { userId: user.id } as ICurrentUser;
+
+		authorizationService.getUserWithPermissions.mockResolvedValue(user);
+		externalToolService.isNameUnique.mockResolvedValue(true);
+		externalToolService.isClientIdUnique.mockResolvedValue(true);
+		externalToolService.hasDuplicateAttributes.mockReturnValue(false);
+		externalToolService.validateByRegex.mockReturnValue(true);
+		externalToolService.createExternalTool.mockResolvedValue(externalToolDO);
+		oauthProviderService.createOAuth2Client.mockResolvedValue(oauthClient);
+		externalToolMapper.mapDoToProviderOauthClient.mockReturnValue(oauthClient);
+		externalToolMapper.applyProviderOauthClientToDO.mockReturnValue(oauth2ConfigWithExternalData);
+
+		return {
+			externalToolDO,
```

Comment:
Ceterum censeo Carthaginem esse delendam

Reply:
I think oauthSetup above the describes is ok,  because it is used in two describes.